File: src/pymarktools/core/refactor.py
"""Python compatibility facade for native Markdown refactoring."""

from dataclasses import dataclass
from pathlib import Path

from pymarktools import _native


@dataclass
class FileReference:
    """Information about a file reference found in Markdown."""

    file_path: Path
    line_number: int
    reference_text: str
    reference_type: str
    target_path: str


class FileReferenceManager:
    """Manage Markdown references through the Rust core."""

    def __init__(self, base_dir: Path) -> None:
        self.base_dir = Path(base_dir).resolve()

    def find_references(
        self,
        target_file: Path,
        include_pattern: str = "*.md",
        exclude_pattern: str | None = None,
    ) -> list[FileReference]:
        """Find all references to ``target_file`` within Markdown files."""
        return [
            FileReference(
                file_path=Path(reference["file_path"]),
                line_number=int(reference["line_number"]),
                reference_text=str(reference["reference_text"]),
                reference_type=str(reference["reference_type"]),
                target_path=str(reference["target_path"]),
            )
            for reference in _native.find_references(
                str(target_file),
                str(self.base_dir),
                include_pattern,
                exclude_pattern,
            )
        ]

    def _get_search_files(self, include_pattern: str, exclude_pattern: str | None) -> list[Path]:
        """Return Markdown files selected by the native discoverer."""
        return [
            Path(path)
            for path in _native.discover_files(
                str(self.base_dir),
                include_pattern,
                exclude_pattern,
                False,
            )
        ]

    def _is_reference_to_target(self, ref_path: str, target_file: Path, source_file: Path) -> bool:
        """Return whether a Markdown target resolves to ``target_file``."""
        if ref_path.startswith(("http://", "https://")):
            return False
        try:
            resolved = (
                self.base_dir / ref_path.lstrip("/")
                if ref_path.startswith("/")
                else Path(_native.resolve_local_path(ref_path, str(source_file)))
            )
            return resolved.resolve() == target_file.resolve()
        except OSError:
            return False

    def _calculate_relative_path(self, from_dir: Path, to_file: Path) -> str:
        """Calculate a portable Markdown path through the native core."""
        return _native.relative_reference(str(from_dir), str(to_file))

    def calculate_new_reference(self, reference: FileReference, old_path: Path, new_path: Path) -> str:
        """Calculate the replacement Markdown reference after a move."""
        new_relative_path = self._calculate_relative_path(reference.file_path.parent, new_path)
        if reference.reference_type in {"link", "image"}:
            return _native.rewrite_reference(reference.reference_text, reference.target_path, new_relative_path)
        return new_relative_path

    def move_file_and_update_references(
        self,
        source: Path,
        destination: Path,
        references: list[FileReference],
        include_pattern: str = "*.md",
        exclude_pattern: str | None = None,
    ) -> None:
        """Move ``source`` and rewrite Markdown references through Rust."""
        del references
        _native.move_and_rewrite(
            str(source),
            str(destination),
            str(self.base_dir),
            include_pattern,
            exclude_pattern,
        )

    def _update_file_content(
        self,
        content: str,
        references: list[FileReference],
        old_path: Path,
        new_path: Path,
    ) -> str:
        """Compatibility helper for replacing known references in text."""
        lines = content.split("\n")
        for reference in sorted(references, key=lambda value: value.line_number, reverse=True):
            if Path(reference.target_path).name != old_path.name:
                continue
            line_index = reference.line_number - 1
            if 0 <= line_index < len(lines):
                replacement = self.calculate_new_reference(reference, old_path, new_path)
                lines[line_index] = lines[line_index].replace(reference.reference_text, replacement)
        return "\n".join(lines)

File: crates/pymarktools-core/src/refactor.rs
//! Reference path calculations used by Markdown refactoring.

use std::fs;
use std::path::{Component, Path, PathBuf};

use regex::Regex;

/// A Markdown reference to a file discovered during refactoring.
#[derive(Debug, Eq, PartialEq)]
pub struct FileReference {
    /// Markdown file containing the reference.
    pub file_path: PathBuf,
    /// One-based source line number.
    pub line_number: usize,
    /// Complete Markdown reference text.
    pub reference_text: String,
    /// Either `link` or `image`.
    pub reference_type: String,
    /// The original target text inside the Markdown reference.
    pub target_path: String,
}

/// Calculate a forward-slash relative reference from one directory to a file.
#[must_use]
pub fn relative_reference(from_dir: &Path, to_file: &Path) -> String {
    let from = components(from_dir);
    let to = components(to_file);
    let common = from
        .iter()
        .zip(&to)
        .take_while(|(left, right)| left == right)
        .count();
    let mut parts = vec!["..".to_owned(); from.len().saturating_sub(common)];
    parts.extend(to.into_iter().skip(common));
    if parts.is_empty() {
        ".".to_owned()
    } else {
        parts.join("/")
    }
}

/// Replace one target path inside a Markdown link or image reference.
#[must_use]
pub fn rewrite_reference(reference: &str, old_target: &str, new_target: &str) -> String {
    reference.replacen(&format!("({old_target})"), &format!("({new_target})"), 1)
}

/// Find Markdown links and images that resolve to `target_file`.
pub fn find_references(
    target_file: &Path,
    base_dir: &Path,
    include_pattern: &str,
    exclude_pattern: Option<&str>,
) -> Result<Vec<FileReference>, String> {
    let target = target_file
        .canonicalize()
        .map_err(|error| error.to_string())?;
    let files = crate::discovery::discover_markdown_files(
        base_dir,
        include_pattern,
        exclude_pattern,
        false,
    )?;
    let pattern = Regex::new(r"(!?)\[[^\]]*\]\(([^)]+)\)").map_err(|error| error.to_string())?;
    let mut references = Vec::new();

    for file_path in files {
        let content = fs::read_to_string(&file_path).map_err(|error| error.to_string())?;
        for (line_index, line) in content.lines().enumerate() {
            for captures in pattern.captures_iter(line) {
                let target_path = captures[2].to_owned();
                if target_path.starts_with("http://") || target_path.starts_with("https://") {
                    continue;
                }
                let candidate = resolve_reference_target(&target_path, &file_path, base_dir);
                if candidate
                    .canonicalize()
                    .as_ref()
                    .is_ok_and(|path| path == &target)
                {
                    references.push(FileReference {
                        file_path: file_path.clone(),
                        line_number: line_index + 1,
                        reference_text: captures[0].to_owned(),
                        reference_type: if captures[1].is_empty() {
                            "link".to_owned()
                        } else {
                            "image".to_owned()
                        },
                        target_path,
                    });
                }
            }
        }
    }

    Ok(references)
}

/// Move a file and update Markdown links and images that reference it.
pub fn move_and_rewrite(
    source: &Path,
    destination: &Path,
    base_dir: &Path,
    include_pattern: &str,
    exclude_pattern: Option<&str>,
    selected_files: Option<&[PathBuf]>,
) -> Result<(), String> {
    let source = source.canonicalize().map_err(|error| error.to_string())?;
    let base_dir = base_dir.canonicalize().map_err(|error| error.to_string())?;
    let destination_parent = destination
        .parent()
        .filter(|path| !path.as_os_str().is_empty())
        .unwrap_or_else(|| Path::new("."));
    fs::create_dir_all(destination_parent).map_err(|error| error.to_string())?;
    let destination_name = destination
        .file_name()
        .ok_or_else(|| "destination must name a file".to_owned())?;
    let destination = destination_parent
        .canonicalize()
        .map_err(|error| error.to_string())?
        .join(destination_name);
    let files = match selected_files {
        Some(files) => files
            .iter()
            .map(|path| path.canonicalize().map_err(|error| error.to_string()))
            .collect::<Result<Vec<_>, _>>()?,
        None => crate::discovery::discover_markdown_files(
            &base_dir,
            include_pattern,
            exclude_pattern,
            false,
        )?,
    };
    let pattern = Regex::new(r"!?\[[^\]]*\]\(([^)]+)\)").map_err(|error| error.to_string())?;
    let mut changes = Vec::<(PathBuf, String)>::new();
    for file in files {
        let content = fs::read_to_string(&file).map_err(|error| error.to_string())?;
        let replacement = relative_reference(file.parent().unwrap_or(&base_dir), &destination);
        let updated = pattern
            .replace_all(&content, |captures: &regex::Captures<'_>| {
                let target = &captures[1];
                if target.starts_with("http://") || target.starts_with("https://") {
                    return captures[0].to_owned();
                }
                let resolved = resolve_reference_target(target, &file, &base_dir).canonicalize();
                if resolved.as_ref().is_ok_and(|path| path == &source) {
                    rewrite_reference(&captures[0], target, &replacement)
                } else {
                    captures[0].to_owned()
                }
            })
            .into_owned();
        if updated != content {
            changes.push((file, updated));
        }
    }
    fs::rename(&source, destination).map_err(|error| error.to_string())?;
    for (file, content) in changes {
        fs::write(file, content).map_err(|error| error.to_string())?;
    }
    Ok(())
}

fn resolve_reference_target(target: &str, source_file: &Path, base_dir: &Path) -> PathBuf {
    if target.starts_with('/') {
        base_dir.join(target.trim_start_matches('/'))
    } else {
        source_file.parent().unwrap_or(base_dir).join(target)
    }
}

fn components(path: &Path) -> Vec<String> {
    path.components()
        .filter_map(|component| match component {
            Component::Normal(value) => Some(value.to_string_lossy().into_owned()),
            _ => None,
        })
        .collect()
}
